state/megawatcher.go

```diff
@@ -491,6 +496,86 @@ func backingEntityIdForSettingsKey(key string) (eid multiwatcher.EntityId, extra
 	return
 }
 
+type backingOpenedPorts map[string]interface{}
+
+func (p *backingOpenedPorts) updated(st *State, store *multiwatcherStore, id interface{}) error {
+	localID := st.localID(id.(string))
+	parentId, ok := backingEntityIdForOpenedPortsKey(localID)
+	if !ok {
+		return nil
+	}
+	info0 := store.Get(parentId)
+	switch info := info0.(type) {
+	case nil:
+		// The parent info doesn't exist. This is unexpected because the port
+		// always refers to a machine. Anyway, ignore the ports for now.
+		return nil
+	case *multiwatcher.MachineInfo:
+		// Retrieve the machine.
+		m, err := st.Machine(info.Id)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		// Update the ports on all units assigned to the machine.
+		units, err := m.Units()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, u := range units {
+			if err := updateUnitPorts(st, store, u); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	default:
+		return nil
+	}
+	return nil
+}
+
+func (p *backingOpenedPorts) removed(st *State, store *multiwatcherStore, id interface{}) {}
+
+func (p *backingOpenedPorts) mongoId() interface{} {
+	panic("cannot find mongo id from openedPorts document")
+}
+
+// updateUnitPorts updates the Ports and PortRanges info of the given unit.
+func updateUnitPorts(st *State, store *multiwatcherStore, u *Unit) error {
+	eid, ok := backingEntityIdForGlobalKey(u.globalKey())
+	if !ok {
+		// This should never happen.
+		return errors.New("cannot retrieve entity id for unit")
+	}
+	info0 := store.Get(eid)
+	switch oldInfo := info0.(type) {
+	case nil:
+		// The unit info doesn't exist. This is unlikely to happen, but ignore
+		// the status until a unitInfo is included in the store.
+		return nil
+	case *multiwatcher.UnitInfo:
+		portRanges, compatiblePorts, err := getUnitPortRangesAndPorts(st, oldInfo.Name)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		unitInfo := *oldInfo
+		unitInfo.PortRanges = portRanges
+		unitInfo.Ports = compatiblePorts
+		store.Update(&unitInfo)
+	default:
+		return nil
+	}
+	return nil
+}
+
+// backingEntityIdForOpenedPortsKey returns the entity id for the given
+// openedPorts key. Any extra information in the key is discarded.
+func backingEntityIdForOpenedPortsKey(key string) (multiwatcher.EntityId, bool) {
+	i := strings.Index(key, "#n")
```

Comment:
Since you're in the state package, how about this instead?

```go
machineId, err := extractPortsIdParts(key)
if err != nil {
    logger.Debugf("cannot parse ports key %q: %v", key, err)
    return multiwatcher.EntityId{}, false
}
return backingEntityIdForGlobalKey(machineGlobalKey(machineId))
```